rclpy/rclpy/executors.py

```diff
@@ -305,93 +291,73 @@ def wait_for_ready_callbacks(self, timeout_sec=None, nodes=None):
 
             # Construct a wait set
             with _WaitSet() as wait_set:
-                _rclpy.rclpy_wait_set_init(
-                    wait_set,
-                    len(subscriptions),
-                    len(guards) + 2,
-                    len(timers),
-                    len(clients),
-                    len(services))
-
-                entities = {
-                    'subscription': (subscriptions, 'subscription_handle'),
-                    'guard_condition': (guards, 'guard_handle'),
-                    'client': (clients, 'client_handle'),
-                    'service': (services, 'service_handle'),
-                    'timer': (timers, 'timer_handle'),
-                }
-                for entity, (handles, handle_name) in entities.items():
-                    _rclpy.rclpy_wait_set_clear_entities(entity, wait_set)
-                    for h in handles:
-                        _rclpy.rclpy_wait_set_add_entity(
-                            entity, wait_set, h.__getattribute__(handle_name)
-                        )
-                _rclpy.rclpy_wait_set_add_entity('guard_condition', wait_set, sigint_gc)
-                _rclpy.rclpy_wait_set_add_entity(
-                    'guard_condition', wait_set, self._guard_condition)
+                wait_set.add_subscriptions(subscriptions)
+                wait_set.add_clients(clients)
+                wait_set.add_services(services)
+                wait_set.add_timers(timers)
+                wait_set.add_guard_conditions(guards)
+                wait_set.add_guard_condition(sigint_gc, sigint_gc_handle)
+                wait_set.add_guard_condition(self._guard_condition, self._guard_condition_handle)
 
                 # Wait for something to become ready
-                _rclpy.rclpy_wait(wait_set, timeout_nsec)
-
-                # get ready entities
-                subs_ready = _rclpy.rclpy_get_ready_entities('subscription', wait_set)
-                guards_ready = _rclpy.rclpy_get_ready_entities('guard_condition', wait_set)
-                timers_ready = _rclpy.rclpy_get_ready_entities('timer', wait_set)
-                clients_ready = _rclpy.rclpy_get_ready_entities('client', wait_set)
-                services_ready = _rclpy.rclpy_get_ready_entities('service', wait_set)
+                wait_set.wait(timeout_nsec)
+
+                # Check sigint guard condition
+                if wait_set.is_ready(sigint_gc_handle):
+                    raise KeyboardInterrupt
+                _rclpy.rclpy_destroy_entity(sigint_gc)
+
+                # Mark all guards as triggered before yielding since they're auto-taken
+                for gc in [g for g in guards if wait_set.is_ready(g.guard_pointer)]:
```

Comment:
This doesn't need to be changed in this PR since we do it "everywhere" in the code base like this.

I was just wondering if you should migrate to use something like this instead in the future:

```
for gc in filter(lambda g: wait_set.is_ready(g.guard_pointer), guards):
```

Just curious what others think...?